ts/src/pro/poloniexfutures.ts

```diff
@@ -0,0 +1,816 @@
+         */
+        await this.loadMarkets ();
+        const name = '/contractMarket/execution';
+        symbol = this.symbol (symbol);
+        const trades = await this.subscribe (name, false, symbol, params);
+        if (this.newUpdates) {
+            limit = trades.getLimit (symbol, limit);
+        }
+        return this.filterBySinceLimit (trades, since, limit, 'timestamp', true);
+    }
+
+    async watchOrderBook (symbol: string, limit: Int = undefined, params = {}) {
+        /**
+         * @method
+         * @name poloniexfutures#watchOrderBook
+         * @description watches information on open orders with bid (buy) and ask (sell) prices, volumes and other data
+         * @see https://futures-docs.poloniex.com/#level-2-market-data
+         * @param {string} symbol unified symbol of the market to fetch the order book for
+         * @param {int|undefined} limit not used by poloniexfutures watchOrderBook
+         * @param {object} params extra parameters specific to the poloniexfutures api endpoint
+         * @returns {object} A dictionary of [order book structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-book-structure} indexed by market symbols
+         */
+        const options = this.safeValue (this.options, 'watchOrderBook');
+        let name = this.safeString (options, 'method', '/contractMarket/level3v2'); // can also be /contractMarket/level2
+        [ name, params ] = this.handleOptionAndParams (params, 'method', 'name', name);
+        const orderbook = await this.subscribe (name, false, symbol, params);
+        return orderbook.limit ();
+    }
+
+    async watchOrders (symbol: string = undefined, since: Int = undefined, limit: Int = undefined, params = {}) {
+        /**
+         * @method
+         * @name poloniexfutures#watchOrders
+         * @description watches information on multiple orders made by the user
+         * @see https://futures-docs.poloniex.com/#private-messages
+         * @param {string|undefined} symbol not used by poloniexfutures watchOrders
+         * @param {int|undefined} since not used by poloniexfutures watchOrders
+         * @param {int|undefined} limit not used by poloniexfutures watchOrders
+         * @param {object} params extra parameters specific to the poloniexfutures api endpoint
+         * @returns {[object]} a list of [order structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-structure}
+         */
+        await this.loadMarkets ();
+        const stop = this.safeValue (params, 'stop');
+        const name = stop ? '/contractMarket/advancedOrders' : '/contractMarket/tradeOrders';
+        const orders = await this.subscribe (name, true, undefined, params);
+        if (this.newUpdates) {
+            limit = orders.getLimit (symbol, limit);
+        }
+        return this.filterBySinceLimit (orders, since, limit, 'timestamp', true);
+    }
+
+    async watchBalance (params = {}) {
```

Comment:
I get the following error in python:
```
  File "/home/pablo/github/ccxt/python/ccxt/pro/poloniexfutures.py", line 109, in subscribe
    'id': milliseconds + name + symbol,   # ID is a unique string to mark the request which is same id property of ack.
TypeError: can only concatenate str (not "NoneType") to str

```

Reply:
@pcriadoperez you can commit directly to this branch to try and fix the errors before I give it a final review in a few weeks